File: solution/3400-3499/3413.Maximum Coins From K Consecutive Bags/Solution.py
class Solution:
    def maximumCoins(self, coins: List[List[int]], k: int) -> int:
        coins.sort(key=itemgetter(0))
        i, total, ans, n = 0, 0, 0, len(coins)
        l0, r0, c0 = coins[0]
        print(coins)
        for l, r, c in coins:
            while i < n and r - coins[i][0] + 1 >= k:
                # print(r,coins[i][0],total)
                l0, r0, c0 = coins[i]
                ans = max(total + max(k + l0 - l, 0) * c, ans)
                total -= (r0 - l0 + 1) * c0
                i += 1

            total += (r - l + 1) * c
            # print(total,l,r)
            if i < n and coins[i][0] != l0:
                start_l = r - k + 1
                ans = max(total + max(r0 - start_l + 1, 0) * c0, ans)
        return max(ans, total)

File: solution/3400-3499/3413.Maximum Coins From K Consecutive Bags/Solution.go
func maximumCoins(coins [][]int, k int) int64 {
	sort.Slice(coins, func(i, j int) bool {
		return coins[i][0] < coins[j][0]
	})

	n := len(coins)
	var total int64 = 0
	var ans int64 = 0
	i := 0
	l0, r0, c0 := coins[0][0], coins[0][1], coins[0][2]

	for _, seg := range coins {
		l, r, c := seg[0], seg[1], seg[2]

		for i < n && r-coins[i][0]+1 >= k {
			l0, r0, c0 = coins[i][0], coins[i][1], coins[i][2]
			ans = max64(ans, total+int64(max(k+l0-l, 0))*int64(c))
			total -= int64(r0-l0+1) * int64(c0)
			i++
		}

		total += int64(r-l+1) * int64(c)

		if i < n && coins[i][0] != l0 {
			startL := r - k + 1
			ans = max64(ans, total+int64(max(r0-startL+1, 0))*int64(c0))
		}
	}
	if total > ans {
		ans = total
	}
	return ans
}

func max64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
